components/brave_wallet/browser/cardano/cardano_api_impl.cc

```diff
@@ -47,6 +48,101 @@ constexpr char kNotImplemented[] = "Not implemented";
 [[maybe_unused]] constexpr int kTxSignProofGeneration = 1;
 [[maybe_unused]] constexpr int kTxSignUserDeclined = 2;
 
+base::expected<
+    std::vector<std::pair<CardanoAddress, cardano_rpc::UnspentOutput>>,
+    mojom::CardanoProviderErrorBundlePtr>
+FilterUtxosByAmount(const GetCardanoUtxosTask::UtxoMap& utxo_map,
+                    const std::optional<std::string>& amount) {
+  std::vector<std::pair<CardanoAddress, cardano_rpc::UnspentOutput>> utxos;
+  if (amount) {
+    auto numeric_amount = CardanoCip30Serializer::DeserializeAmount(*amount);
+
+    if (!numeric_amount) {
+      return base::unexpected(mojom::CardanoProviderErrorBundle::New(
+          kAPIErrorInternalError, "Failed to decode amount", nullptr));
+    }
+
+    // TODO(cypt4): Use knapsack solver here
+    base::CheckedNumeric<uint64_t> accumulated_sum = 0u;
+
+    for (const auto& pair_by_address : utxo_map) {
+      for (const auto& utxo : pair_by_address.second) {
+        accumulated_sum += utxo.lovelace_amount;
+        if (!accumulated_sum.IsValid()) {
+          return base::unexpected(mojom::CardanoProviderErrorBundle::New(
+              kAPIErrorInternalError, "Value overflow", nullptr));
+        }
+
+        utxos.push_back({pair_by_address.first, utxo});
+
+        if (accumulated_sum.ValueOrDie() >= numeric_amount.value()) {
+          break;
+        }
+      }
+    }
+
+    if (accumulated_sum.ValueOrDie() < numeric_amount.value()) {
+      return base::unexpected(mojom::CardanoProviderErrorBundle::New(
+          kAPIErrorInternalError, "Not enough funds", nullptr));
+    }
+  } else {
+    for (const auto& pair_by_address : utxo_map) {
+      for (const auto& utxo : pair_by_address.second) {
+        utxos.push_back({pair_by_address.first, utxo});
+      }
+    }
```

Comment:
Let's do this first. Then sort by tx_hash/output_index, so utxos in some predictable order. 
Then return either all these utxos, or(in case of amount arg) some of them starting from the front.